actors/builtin/power/power_test.go

```diff
@@ -752,6 +765,161 @@ func TestSubmitPoRepForBulkVerify(t *testing.T) {
+			{miner3, []int{200, 201}},
+			{miner4, []int{300, 301}},
+			{miner2, []int{100, 101}}}
+
+		infos := map[addr.Address][]abi.SealVerifyInfo{miner1: []abi.SealVerifyInfo{*info1, *info2},
+			miner2: []abi.SealVerifyInfo{*info3, *info4},
+			miner3: []abi.SealVerifyInfo{*info5, *info6},
+			miner4: []abi.SealVerifyInfo{*info7, *info8}}
+
+		ac.createMinerBasic(rt, owner, owner, miner1)
+		ac.createMinerBasic(rt, owner, owner, miner2)
+		ac.createMinerBasic(rt, owner, owner, miner3)
+		ac.createMinerBasic(rt, owner, owner, miner4)
+		ac.updateClaimedPower(rt, miner1, powerUnit, powerUnit)
+		ac.updateClaimedPower(rt, miner1, powerUnit, powerUnit)
+		ac.updateClaimedPower(rt, miner1, powerUnit, powerUnit)
+		ac.updateClaimedPower(rt, miner1, powerUnit, powerUnit)
+
+		expectedPower := big.Mul(big.NewInt(4), powerUnit)
+
+		delta := abi.NewTokenAmount(1)
+		ac.updatePledgeTotal(rt, miner1, delta)
+		ac.onEpochTickEnd(rt, 0, expectedPower, cs, infos)
+
+		st := getState(rt)
+		require.EqualValues(t, delta, st.ThisEpochPledgeCollateral)
+		require.EqualValues(t, expectedPower, st.ThisEpochQualityAdjPower)
+		require.EqualValues(t, expectedPower, st.ThisEpochRawBytePower)
+	})
+
+	t.Run("success when no confirmed sector", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		ac.onEpochTickEnd(rt, 0, big.Zero(), nil, nil)
+	})
+
+	t.Run("fails if batch verify seals does not return result for a miner", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		miner2 := tutil.NewIDAddr(t, 201)
+		info1 := sealInfo(1)
+		info2 := sealInfo(2)
+
+		ac.submitPoRepForBulkVerify(rt, miner1, info1)
+		ac.submitPoRepForBulkVerify(rt, miner2, info2)
+
+		infos := map[addr.Address][]abi.SealVerifyInfo{miner1: []abi.SealVerifyInfo{*info1},
+			miner2: []abi.SealVerifyInfo{*info2}}
+
+		res := batchVerifyDefaultOutput(infos)
+		delete(res, miner1)
+		rt.ExpectBatchVerifySeals(infos, res, nil)
+		rt.ExpectValidateCallerAddr(builtin.CronActorAddr)
+
+		rt.SetEpoch(abi.ChainEpoch(0))
+		rt.SetCaller(builtin.CronActorAddr, builtin.CronActorCodeID)
+
+		rt.ExpectAbort(exitcode.ErrNotFound, func() {
+			rt.Call(ac.Actor.OnEpochTickEnd, nil)
+		})
+		rt.Verify()
+	})
+
+	t.Run("fails if batch verify seals fails", func(t *testing.T) {
+		rt, ac := basicPowerSetup(t)
+		info1 := sealInfo(1)
+		info2 := sealInfo(2)
+		info3 := sealInfo(3)
+
+		ac.submitPoRepForBulkVerify(rt, miner1, info1)
+		ac.submitPoRepForBulkVerify(rt, miner1, info2)
+		ac.submitPoRepForBulkVerify(rt, miner1, info3)
+
+		infos := map[addr.Address][]abi.SealVerifyInfo{miner1: []abi.SealVerifyInfo{*info1, *info2, *info3}}
+
+		rt.ExpectBatchVerifySeals(infos, batchVerifyDefaultOutput(infos), fmt.Errorf("fail"))
+		rt.ExpectValidateCallerAddr(builtin.CronActorAddr)
+
+		rt.SetEpoch(abi.ChainEpoch(0))
+		rt.SetCaller(builtin.CronActorAddr, builtin.CronActorCodeID)
+
+		rt.ExpectAbort(exitcode.ErrIllegalState, func() {
+			rt.Call(ac.Actor.OnEpochTickEnd, nil)
+		})
+		rt.Verify()
+	})
```

Comment:
Please test
- duplicate SVIs for same sector (dup ignored)
- failed verification for one SVI for a miner, with others that succeed

Reply:
@anorth The test for duplicates is done.

However, I'll write a test for the failure case once we agree on https://github.com/filecoin-project/specs-actors/pull/831#discussion_r462261522.

Reply:
Maybe I still don't understand, but as noted above I think the semantics are exactly the same either way.

I'm talking about returning a `false` value, not incorrectly failing to return a result for a miner at all.

Reply:
This is done.